src/lib/SoftHSM.cpp

```diff
@@ -11489,23 +11524,76 @@ CK_RV SoftHSM::deriveSymmetric
 			bOK = bOK && osobject->setAttribute(CKA_LOCAL,false);
 
 			// Common Secret Key Attributes
-			if (baseKey->getBooleanValue(CKA_ALWAYS_SENSITIVE, false))
-			{
-				bool bAlwaysSensitive = osobject->getBooleanValue(CKA_SENSITIVE, false);
-				bOK = bOK && osobject->setAttribute(CKA_ALWAYS_SENSITIVE,bAlwaysSensitive);
-			}
-			else
-			{
-				bOK = bOK && osobject->setAttribute(CKA_ALWAYS_SENSITIVE,false);
-			}
-			if (baseKey->getBooleanValue(CKA_NEVER_EXTRACTABLE, true))
-			{
-				bool bNeverExtractable = osobject->getBooleanValue(CKA_EXTRACTABLE, false) == false;
-				bOK = bOK && osobject->setAttribute(CKA_NEVER_EXTRACTABLE,bNeverExtractable);
-			}
-			else
-			{
-				bOK = bOK && osobject->setAttribute(CKA_NEVER_EXTRACTABLE,false);
+			if (pMechanism->mechanism == CKM_CONCATENATE_BASE_AND_KEY) {
+				// [PKCS#11 v2.40, 2.31.3]
+				// If either of the two original keys has its CKA_SENSITIVE attribute
+				// set to CK_TRUE, so does the derived key.  If not, then the derived
+				// key’s CKA_SENSITIVE attribute is set either from the supplied template
+				// or from a default value.
+				bool bSensitive = baseKey->getBooleanValue(CKA_SENSITIVE, true) ||
+								  otherKey->getBooleanValue(CKA_SENSITIVE, true);
+				if (bSensitive) {
+					bOK = bOK && osobject->setAttribute(CKA_SENSITIVE, true);
+				}
+				// If either of the two original keys has its CKA_EXTRACTABLE attribute
+				// set to CK_FALSE, so does the derived key.  If not, then the derived
+				// key’s CKA_EXTRACTABLE attribute is set either from the supplied template
+				// or from a default value.
+				bool bExtractable = baseKey->getBooleanValue(CKA_EXTRACTABLE, false) ||
+									otherKey->getBooleanValue(CKA_EXTRACTABLE, false);
+				if (!bExtractable) {
+					bOK = bOK && osobject->setAttribute(CKA_EXTRACTABLE, false);
+				}
+				// The derived key’s CKA_ALWAYS_SENSITIVE attribute is set to CK_TRUE
+				// if and only if both of the original keys have their CKA_ALWAYS_SENSITIVE
+				// attributes set to CK_TRUE.
+				bool bAlwaysSensitive = baseKey->getBooleanValue(CKA_ALWAYS_SENSITIVE, false) &&
+										otherKey->getBooleanValue(CKA_ALWAYS_SENSITIVE, false);
+				bOK = bOK && osobject->setAttribute(CKA_ALWAYS_SENSITIVE, bAlwaysSensitive);
+				// The derived key’s CKA_NEVER_EXTRACTABLE attribute is set to CK_TRUE
+				// if and only if both of the original keys have their CKA_NEVER_EXTRACTABLE
+				// attributes set to CK_TRUE
+				bool bNeverExtractable = baseKey->getBooleanValue(CKA_NEVER_EXTRACTABLE, false) &&
+										 otherKey->getBooleanValue(CKA_NEVER_EXTRACTABLE, false);
+				bOK = bOK && osobject->setAttribute(CKA_ALWAYS_SENSITIVE, bNeverExtractable);
+			}
+			else if (pMechanism->mechanism == CKM_CONCATENATE_BASE_AND_DATA ||
+				 pMechanism->mechanism == CKM_CONCATENATE_DATA_AND_BASE)
```

Comment:
Probably just an irrelevant nitpick, if this PR is accepted as a whole: this branch is added only as a part of the commit introducing `CKM_CONCATENATE_BASE_AND_KEY`.

Reply:
Should I split this commit?

Reply:
Certainly not because of me.